src/services/sanity.js

```diff
@@ -1301,6 +1310,42 @@ export async function fetchSanity(query, isList) {
   }
 }
 
+async function needsAccessDecorator(results)
+{
+    if (globalConfig.sanityConfig.useDummyRailContentMethods) return results;
+    let userPermissions = await getUserPermissions();
+    userPermissions = new Set(userPermissions);
+    if (Array.isArray(results)) {
+        results.forEach((result) => {
+            result['need_access'] = doesUserNeedAccessToContent(result, userPermissions);
+        });
+    } else {
+        results['need_access'] = doesUserNeedAccessToContent(results, userPermissions);
+    }
+    return results;
+}
+
+function doesUserNeedAccessToContent(result, userPermissions)
+{
+    const permissions =  new Set(result.permission_id ?? []);
+    if (permissions.length === 0) {
+        return false;
+    }
+    for (let permission of permissions) {
+        if (userPermissions.has(permission)) {
+            return false;
+        }
+    }
+    return true;
+}
+
+async function getUserPermissions()
+{
+    const permissions = await fetchUserPermissions();
+    return permissions;
```

Comment:
Some more nitpicking, but to be a bit cleaner:
```suggestion
    return await fetchUserPermissions();
```